### PR TITLE
refactor: Move validation logic from services to handlers

### PR DESCRIPTION
**Pull request by VanillaSpoon**
## Description
Move validation logic from services to handlers
[MGDSTRM-6562](https://issues.redhat.com/browse/MGDSTRM-6562)

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

File: internal/kafka/internal/handlers/validation.go [hunk 1]
```diff
@@ -93,6 +97,96 @@ func ValidateKafkaClusterNameIsUnique(name *string, kafkaService services.KafkaS
...
+
+		kafkaVersionAvailable, err2 := h.clusterService.IsStrimziKafkaVersionAvailableInCluster(cluster, desiredStrimziVersion, desiredKafkaVersion, desiredKafkaIBPVersion)
+
+		if err2 != nil {
+			return errors.Validation(err2.Error())
+		}
+
+		if !kafkaVersionAvailable {
+			return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to update kafka: %s with kafka version: %s", kafkaRequest.ID, desiredKafkaVersion))
+		}
```
Comment (ziccardi) [inline internal/kafka/internal/handlers/validation.go:139]:
```suggestion
if kafkaVersionAvailable, err := h.clusterService.IsStrimziKafkaVersionAvailableInCluster(cluster, desiredStrimziVersion, desiredKafkaVersion, desiredKafkaIBPVersion); err !=nil {
    return errors.Validation(err.Error())
} else if !kafkaVersionAvailable {
    return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to update kafka: %s with kafka version: %s", kafkaRequest.ID, desiredKafkaVersion))
}
```

File: internal/kafka/internal/handlers/validation.go [hunk 2]
```diff
@@ -93,6 +97,96 @@ func ValidateKafkaClusterNameIsUnique(name *string, kafkaService services.KafkaS
...
+		vCompKafka, ek := api.CompareSemanticVersionsMajorAndMinor(currentKafkaVersion, desiredKafkaVersion)
+
+		if ek != nil {
+			return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to compare desired kafka version: %s with actual kafka version: %s", desiredKafkaVersion, currentKafkaVersion))
+		}
+
+		// no minor/ major version downgrades allowed for kafka version
+		if vCompKafka > 0 {
+			return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to downgrade kafka: %s version: %s to the following kafka version: %s", kafkaRequest.ID, currentKafkaVersion, desiredKafkaVersion))
+		}
```
Comment (ziccardi) [inline internal/kafka/internal/handlers/validation.go:184]:
```suggestion
		if vCompKafka, err := api.CompareSemanticVersionsMajorAndMinor(currentKafkaVersion, desiredKafkaVersion); err != nil {
			return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to compare desired kafka version: %s with actual kafka version: %s", desiredKafkaVersion, currentKafkaVersion))
		} else if vCompKafka > 0 {
			return errors.New(errors.ErrorValidation, fmt.Sprintf("Unable to downgrade kafka: %s version: %s to the following kafka version: %s", kafkaRequest.ID, currentKafkaVersion, desiredKafkaVersion))
		}
```

**Comment by ziccardi** [comment]
I believe it would be hard to read the code and understand that a variable named `eIbp` or `ek` is an error.
Proposed changes allow to always use `err`.

**Comment by VanillaSpoon** [comment]
> We are replacing a method using a 'validation' framework where each validation function has its own responsibility with a big method performing all the checks at once. I believe this makes the code less maintainable/readable.
> 
> WDYT? ^^ @machi1990 @VanillaSpoon

Thanks for the feedback @ziccardi, I agree the separation will make the code easier to maintain/read. 

This is something I was hoping to address in a follow-up pr, whereby I will 
- Refactor the validations (Splitting kafka, strimzi, and ibp version validations). 
- Add operations to skip validations if all in kafkaUpdateReq versions are not set.
- Skip the individual validation if a particular version is not set.

File: internal/kafka/internal/handlers/admin_kafka.go [hunk 3]
```diff
@@ -123,35 +120,9 @@ func (h adminKafkaHandler) Update(w http.ResponseWriter, r *http.Request) {
...
-					return errors.New(errors.ErrorValidation, "Unable to update strimzi version. Another upgrade is already in progress.")
-				}
-				return nil
-			},
-			func() *errors.ServiceError { // Validate DesiredKafkaIBPVersion
-				if kafkaRequest.DesiredKafkaIBPVersion != kafkaUpdateReq.KafkaIbpVersion && kafkaUpdateReq.KafkaIbpVersion != "" && kafkaRequest.KafkaIBPUpgrading {
-					return errors.New(errors.ErrorValidation, "Unable to update ibp version. Another upgrade is already in progress.")
-				}
-				return nil
-			},
```
[the remark this answers is not in the document]
Comment (machi1990) [reply internal/kafka/internal/handlers/admin_kafka.go:154]: Oh I've just seen them.

I'll propose we keep those as they were i.e as separate validation function.

The new validation method will be from the extracted code from the services that performed the verification. This new validation function, can then be tested separately.

**Comment by machi1990** [comment]
> Refactor the validations (Splitting kafka, strimzi, and ibp version validations).

I know that there are some dependencies between the versions i.e strimzi -> kafka -> ibp version. 
I think the new validator that checks if the upgrade is not breaking i.e the versions compatible can be performed as one validation.
The validator that checked for versions being set etc can be performed individually as suggested in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1134#discussion_r907264300

File: internal/kafka/internal/handlers/admin_kafka.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (VanillaSpoon) [reply internal/kafka/internal/handlers/admin_kafka.go:154]: Perfect :) thanks for the feedback @machi1990

[the remark this answers is not in the document]
Comment (machi1990) [reply internal/kafka/internal/handlers/admin_kafka.go]: Thanks @VanillaSpoon

**Review by machi1990 — APPROVED** [review]
lgtm, thanks @VanillaSpoon for addressing the feedback and for driving this baby home :-)

I'll leave it to @ziccardi and @pawelpaszki for the final review and approval

File: internal/kafka/internal/handlers/admin_kafka_test.go [hunk 4]
```diff
@@ -244,40 +246,53 @@ func Test_Update(t *testing.T) {
...
+								ID: "id",
+							},
+							ClusterID:              "cluster-id",
+							ActualKafkaIBPVersion:  "2.7",
+							DesiredKafkaIBPVersion: "2.7",
+							ActualKafkaVersion:     "2.7",
+							DesiredKafkaVersion:    "2.7",
+							DesiredStrimziVersion:  "2.7",
+							KafkaStorageSize:       "100",
+						}, errors.GeneralError("test")
```
Comment (ziccardi) [inline internal/kafka/internal/handlers/admin_kafka_test.go:274]: Why returning both a return value and an error? Can't we just return `nil, errors.GeneralError("test")`?

Before this wasn't returning any error on `kafka no found` (was just returning `nil`): is the behaviour changed?

Comment (VanillaSpoon) [reply internal/kafka/internal/handlers/admin_kafka_test.go:274]: Thanks for pointing this out :) no, it was just the case that I copied the kafka request from one of my other tests. I'll switch it out in the morning.

**Comment by pawelpaszki** [comment]
Looks good, once the unresolved Massimo's comment is addressed. Great work!

File: internal/kafka/internal/handlers/admin_kafka_test.go (hunk 4 above)
Comment (VanillaSpoon) [reply internal/kafka/internal/handlers/admin_kafka_test.go]: Thanks @ziccardi, this has been resolved in the latest push :)

**Review by ziccardi — APPROVED** [review]
LGTM. Great job!

**Pull request merged**